### PR TITLE
Project visibility configuration

### PR DESCRIPTION
**Pull request by hlbmtc**
**Changes:**
- Replaced `Project.unlisted` and `Project.add_posts_to_main_feed` with a new enum field: `Project.visibility`.
- Added data migration to backfill `Project.visibility` for existing projects.
- Added the `Project.visibility` field.
- Introduced a deprecation message for the `unlisted` and `add_posts_to_main_feed` fields.
- Updated logic to hide unlisted tournaments from the tournaments page.
- Updated logic to hide unlisted communities and projects from search engines.
- Removed the unused tournament visibility toggle from the tournament page.
- Adjusted community settings to align with the new `Project.visibility` backend changes.

> [!NOTE]
> The `unlisted` and `add_posts_to_main_feed` fields are still referenced by the current running code. If we were to drop them immediately (by removing the fields and the code that uses them at the same time as applying database migrations), there would be a brief period where the updated database schema no longer matches the running code. During that window, any requests handled by old code still expecting these fields could cause errors.
>
> To avoid this, we're taking a two-step approach:
> 1. **First release:**  
>    - Add `Project.visibility` and migrate the schema.
>    - Keep `unlisted` and `add_posts_to_main_feed` in place (but mark them as deprecated), allowing both old and new > code paths to function without errors.
>   
> 2. **Subsequent release(s):**  
>    - Once the new `Project.visibility` logic is fully live and the old code is no longer in use, safely remove the deprecated > fields.

deprecation ticket: #1715 
closes: #1508

### CODE REVIEW

**Review by ncarazon — APPROVED** [review]
LGTM from a FE perspective

File: posts/models.py [hunk 1]
```diff
@@ -359,10 +359,11 @@ def filter_for_main_feed(self):
         """
 
         return self.filter(
-            Q(default_project__add_posts_to_main_feed=True)
+            Q(default_project__visibility=Project.Visibility.NORMAL)
             | Exists(
                 Post.projects.through.objects.filter(
-                    post_id=OuterRef("pk"), project__add_posts_to_main_feed=True
+                    post_id=OuterRef("pk"),
+                    project__visibility=Project.Visibility.NORMAL,
```
Comment (lsabor) [inline posts/models.py:366]: See comment on `projects/models.py`
Just want to make sure accidental setting of visibility on Tags and other non-question project types don't screw this up.

File: projects/migrations/0009_project_visibility.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Generated by Django 5.1.4 on 2024-12-13 14:09
+
+from django.db import migrations, models
+from django.db.models import When, Case, Value, Q
+
+
+def migrate(apps, schema_editor):
+    """
+    We assume that all existing users with an activated email have completed onboarding.
+    """
```
Comment (lsabor) [inline projects/migrations/0009_project_visibility.py:10]: This comment is an artifact from copying another custom migration.
Pls delete

File: projects/migrations/0009_project_visibility.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# Generated by Django 5.1.4 on 2024-12-13 14:09
+
+from django.db import migrations, models
+from django.db.models import When, Case, Value, Q
+
+
+def migrate(apps, schema_editor):
```
Comment (lsabor) [inline projects/migrations/0009_project_visibility.py:7]: not a fan of using the name "migrate" here. I'd prefer a more detailed name like "set_visibility_field"

File: projects/models.py [hunk 4]
```diff
@@ -235,16 +238,22 @@ class SectionTypes(models.TextChoices):
...
     )
     override_permissions = models.ManyToManyField(User, through="ProjectUserPermission")
-    # Not discoverable, but can still be accessed via the URL
-    unlisted = BooleanField(default=False, db_index=True)
+
+    visibility = models.CharField(
+        choices=Visibility.choices,
+        default=Visibility.NOT_IN_MAIN_FEED,
+        db_index=True,
+    )
```
Comment (lsabor) [inline projects/models.py:246]: I'm not sure if I'm a fan of having a default of not in main feed.
Tags, for instance, seem like they should have no visibility setting. Having visibility settable for Tags seems like it would intuivitely do something.
Can we instead have a `Visibility.NOT_APPLICABLE = ""` choice or make this field blank by default?

Also, help text would be good to add while we're here.
Maybe:
```
help_text="""Sets the visibility of this project.
    'Normal' is visible on the main feed, contributes to global leaderboards/medals, and lists the project in the tournaments/question series page. 
    'not_in_main_feed' is not visible in the main feed, but can be searched for, doesn't contribute to global leaderboards/medals, and lists the project in the tournaments/question series page.
    'unlisted' is not visible in the main feed, not searchable, and doesn't contribute to global leaderboards/medals. 

    If this project is not of type "site_main", "tournament", or "question_series", this field should be left blank ''
"""
```

File: projects/models.py [hunk 5]
```diff
@@ -235,16 +238,22 @@ class SectionTypes(models.TextChoices):
...
     # Whether we should display tournament on the homepage
     show_on_homepage = models.BooleanField(default=False, db_index=True)
 
     objects = models.Manager.from_queryset(ProjectsQuerySet)()
 
     # Annotated fields
-    followers_count = models.PositiveIntegerField(default=0, db_index=True, editable=False)
+    followers_count = models.PositiveIntegerField(
+        default=0, db_index=True, editable=False
+    )
```
Comment (lsabor) [inline projects/models.py:256]: I don't have an objection to this, but seems out of scope for this PR

File: projects/services/communities.py [hunk 6]
```diff
@@ -31,7 +31,10 @@ def get_communities_feed(
 
     # If hidden from the main feed
     if unlisted is not None:
-        qs = qs.filter(unlisted=unlisted)
+        if unlisted:
+            qs = qs.filter(visibility=Project.Visibility.UNLISTED)
+        else:
+            qs = qs.exclude(visibility=Project.Visibility.UNLISTED)
```
Comment (lsabor) [inline projects/services/communities.py:37]: Worth double checking that we aren't missing how "not in main feed" should be slid into this.

File: questions/services.py [hunk 7]
```diff
@@ -567,7 +567,8 @@ def update_leaderboards_for_question(question: Question):
             update_global_leaderboards = True
             continue
         update_global_leaderboards = (
-            update_global_leaderboards or project.add_posts_to_main_feed
+            update_global_leaderboards
+            or project.visibility == Project.Visibility.NORMAL
```
Comment (lsabor) [inline questions/services.py:571]: nit: You can move 
```
        update_global_leaderboards = (
            update_global_leaderboards
            or project.visibility == Project.Visibility.NORMAL
```
to above line 565, and then delete 567 since site_main project will have NORMAL visibility.

**Review by lsabor — APPROVED** [review]
Generally looks good to me.

Before merging, I'd like to make sure we're solidified on the help text for the new visibility field and also think through any awkwardness with regards to visibility settings on non-question Project types (e.g. tags).

**Comment by lsabor** [comment]
This isn't for this PR, but we can maybe take advantage of the new `visibility` field to implement personal tags for questions.
Some user could then add tags to posts that they write, which will have a visibility of `unlisted` but have a `created_by` of said user. Then, when they look through their tags, it'll display all tags that either are not "unlisted" or were created by them. You can then create tags directly from the tag section on the question detail page.
When filtering the feed via tag slug, it we will first check if the tag exists and either has a not-unlisted visibility, or is unlisted and has a `created_by` of the logged in user.

I think it's not bad.
This does imply that tags which are public should probably have a visibility of "normal", meaning we'd need to filter for global leaderboard questions by only looking a projects of types "site_main", "question_series", and "tournament", which isn't a hard addition to the query.

File: projects/models.py (hunk 5 above)
Comment (hlbmtc) [reply projects/models.py:256]: It was just autoformatting, no actual changes here

File: projects/models.py (hunk 4 above)
Comment (hlbmtc) [reply projects/models.py:246]: Yeah, I understand your point. I don’t want to introduce extra states either. Right now, whether a project should be shown in the main feed is determined by checking all of its projects, including secondary ones, where tags are typically located. Since tags are still considered projects in this model, I think it makes sense for them to always be set to `NOT_IN_MAIN_FEED` to remain neutral. This aligns naturally with the overall visibility paradigm.

Comment (hlbmtc) [reply projects/models.py]: Added admin construction to force `NOT_IN_MAIN_FEED` for such cases -- https://github.com/Metaculus/metaculus/blob/2c41909d01ad46634c01b5dae5e52471a8014209/projects/admin.py#L249

Comment (lsabor) [reply projects/models.py]: great, no problem.
We can potentially improve (doesn't have to be in this PR) the admin panel to make that field uneditable for Tag instances.
Or, even better, we make Tags / Topics / etc a different entity than Project...

File: projects/models.py (hunk 5 above)
Comment (lsabor) [reply projects/models.py:265]: Oh, haha I should have checked more closely. no problem
